### ISSUE TITLE
Improve calculation of inverse

### ISSUE THREAD

**Issue by TobyBoyne** [TobyBoyne/alfalfa#6]
During the training loop, we calculate the inverse of the covariance matrix, $(K + \sigma_y^2 I)^{-1}$. This is an $O(n^3)$ operation, and is the bottleneck for the training process. This is especially troublesome for MCMC, where we calculate this for each proposal (one per tree + GP hyperparameters) for each step (~ $10^3$ posterior steps), becoming prohibitively expensive.

However, note that the updates to the covariance matrix are always low-rank. For example, pruning a node essentially combines two blocks of the covariance matrix, which is a low rank update.

Can we use this to our advantage?

**Comment by TobyBoyne** [comment]
As a rough benchmark, each term in the tree proposal takes the following amount of time to calculate:
`{'q_ratio': 0.080ms, 'll_ratio': 3.7ms, 'prior_ratio': 0.10ms}`
It is clear that calculating the log likelihood is the bottleneck in this calculation.

These calculations account for about 60-85% of the MCMC runtime (for 10-30 trees). For 30 trees, the runtime is 104 seconds for 500 MCMC steps.

**Comment by TobyBoyne** [comment]
An alternate suggestion would be to make better use of `linear_operators`. This is already a key part of GPyTorch, and implementing the tree kernel using these structures might help automatically identify this sort of structure.
